source/client.c

```diff
@@ -613,6 +623,9 @@ int aws_mqtt_client_connection_set_login(
 
     AWS_LOGF_TRACE(AWS_LS_MQTT_CLIENT, "id=%p: Setting username and password", (void *)connection);
 
+    struct aws_string *pre_username = connection->username;
+    struct aws_string *pre_password = connection->password;
+
     connection->username = aws_string_new_from_array(connection->allocator, username->ptr, username->len);
```

Comment:
I think something similar could be done here with a "create both resources we need and if either fails, just clean them up, otherwise swap them with the real deals, and clean up"